# Explore here

In [14]:
import pandas as pd
import requests
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB
from sklearn.metrics import accuracy_score, classification_report
from sklearn.svm import SVC
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.ensemble import RandomForestClassifier
import joblib
from sklearn.linear_model import LogisticRegression

In [15]:
url = "https://raw.githubusercontent.com/4GeeksAcademy/naive-bayes-project-tutorial/main/playstore_reviews.csv"
df = pd.read_csv(url)
df.to_csv("../data/raw/data.csv", index=False)
df.head()

,package_name,review,polarity
0,com.facebook.katana,privacy at least put some option appear offli...,0
1,com.facebook.katana,"messenger issues ever since the last update, ...",0
2,com.facebook.katana,profile any time my wife or anybody has more ...,0
3,com.facebook.katana,the new features suck for those of us who don...,0
4,com.facebook.katana,forced reload on uploading pic on replying co...,0


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   package_name  891 non-null    object
 1   review        891 non-null    object
 2   polarity      891 non-null    int64 
dtypes: int64(1), object(2)
memory usage: 21.0+ KB


In [17]:
df["package_name"] = df["package_name"].str.strip().str.lower()
df["review"] = df["review"].str.strip().str.lower()
df.columns = ["package_name", "review", "polarity"]

In [18]:
df

,package_name,review,polarity
0,com.facebook.katana,privacy at least put some option appear offlin...,0
1,com.facebook.katana,"messenger issues ever since the last update, i...",0
2,com.facebook.katana,profile any time my wife or anybody has more t...,0
3,com.facebook.katana,the new features suck for those of us who don'...,0
4,com.facebook.katana,forced reload on uploading pic on replying com...,0
...,...,...,...
886,com.rovio.angrybirds,loved it i loooooooooooooovvved it because it ...,1
887,com.rovio.angrybirds,all time legendary game the birthday party lev...,1
888,com.rovio.angrybirds,ads are way to heavy listen to the bad reviews...,0
889,com.rovio.angrybirds,fun works perfectly well. ads aren't as annoyi...,1


In [ ]:
# Separar características y variable objetivo
X = df.drop(columns=["polarity"])  # Variables predictoras
y = df["polarity"]  # Variable objetivo

# Dividir en entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

# Verificar tamaños
print(f"Train: {X_train.shape}, Test: {X_test.shape}")

Train: (712, 2), Test: (179, 2)


In [20]:
# Inicializar CountVectorizer para transformar el texto en datos numéricos
vec_model = CountVectorizer(stop_words="english")

# Aplicar la transformación al conjunto de entrenamiento
X_train_vec = vec_model.fit_transform(X_train["review"]).toarray()

# Aplicar la transformación al conjunto de prueba
X_test_vec = vec_model.transform(X_test["review"]).toarray()

# Ahora ya tienes `X_test_vec` definido y puedes entrenar Naive Bayes


In [21]:
# Modelos
gnb = GaussianNB()
mnb = MultinomialNB()
bnb = BernoulliNB()

# Entrenamiento
gnb.fit(X_train_vec, y_train)
mnb.fit(X_train_vec, y_train)
bnb.fit(X_train_vec, y_train)

# Prediccion
y_pred_gnb = gnb.predict(X_test_vec)
y_pred_mnb = mnb.predict(X_test_vec)
y_pred_bnb = bnb.predict(X_test_vec)

# Evaluar cada modelo
acc_gnb = accuracy_score(y_test, y_pred_gnb)
acc_mnb = accuracy_score(y_test, y_pred_mnb)
acc_bnb = accuracy_score(y_test, y_pred_bnb)

print(f"Precisión GaussianNB: {acc_gnb:.2f}")
print(f"Precisión MultinomialNB: {acc_mnb:.2f}")
print(f"Precisión BernoulliNB: {acc_bnb:.2f}")


Precisión GaussianNB: 0.82
Precisión MultinomialNB: 0.85
Precisión BernoulliNB: 0.78


In [22]:
best_model = MultinomialNB()
best_model.fit(X_train_vec, y_train)

y_pred_best = best_model.predict(X_test_vec)
accuracy_best = accuracy_score(y_test, y_pred_best)

print(f"Precisión del mejor modelo Naive Bayes: {accuracy_best:.2f}")


Precisión del mejor modelo Naive Bayes: 0.85


In [23]:
# Entrenar Random Forest
RF_model = RandomForestClassifier(n_estimators=100, max_depth=10, random_state=42)
RF_model.fit(X_train_vec, y_train)

# Evaluar precisión en test
y_pred_rf = RF_model.predict(X_test_vec)
accuracy_rf = accuracy_score(y_test, y_pred_rf)

print(f"Precisión de Random Forest: {accuracy_rf:.2f}")


Precisión de Random Forest: 0.69


In [24]:
joblib.dump(best_model if accuracy_best > accuracy_rf else RF_model, "optimized_model.pkl")
print("Modelo guardado correctamente.")

Modelo guardado correctamente.


In [25]:
# Entrenar SVM
svm_model = SVC(kernel="linear", C=1.0, random_state=42)
svm_model.fit(X_train_vec, y_train)

# Predicción y evaluación
y_pred_svm = svm_model.predict(X_test_vec)
accuracy_svm = accuracy_score(y_test, y_pred_svm)

print(f"Precisión SVM: {accuracy_svm:.2f}")


Precisión SVM: 0.79


In [ ]:
# Regresión logística
RL_model = LogisticRegression(random_state=42)
RL_model.fit(X_train_vec, y_train)

# Predicción y evaluación
y_pred_RL = RL_model.predict(X_test_vec)
accuracy_RL = accuracy_score(y_test, y_pred_RL)

print(f"Precisión Regresión Logística: {accuracy_RL:.2f}")


Precisión Regresión Logística: 0.83


El mejor modelo luego de estudiarlos es el de Regresion Logistica. Porque tiene mejores resultados el 83% de efectividadad.